source/Ship.cpp

```diff
@@ -2620,7 +2633,168 @@ bool Ship::IsDisabled() const
+}
+
+
+
+// Determine if the ship has any usable weapons.
+bool Ship::IsArmed() const
+{
+	return IsArmed(false);
+}
+
+
+
+// Determine if the ship has any usable weapons against enemies.
+bool Ship::IsArmed(bool includingAntiMissile) const
+{
+	for(const Hardpoint &hardpoint : Weapons())
+	{
+		const Weapon *weapon = hardpoint.GetOutfit();
+		if(weapon && (!hardpoint.IsAntiMissile() || includingAntiMissile))
+		{
+			if(weapon->Ammo() && !OutfitCount(weapon->Ammo()))
+				continue;
+			return true;
+		}
+	}
+	return false;
+}
+
+
+
+// Check for enemies in the in the current system if ship is owned by player.
+bool Ship::IsEnemyInEscortSystem() const
+{
+	if(!IsYours())
+		return false;
+	const Government *gov = (CanBeCarried() && GetParent()) ? GetParent()->GetGovernment() : GetGovernment();
+	const std::vector<std::weak_ptr<Ship>> allEscorts = (CanBeCarried() && GetParent()) ? GetParent()->GetEscorts() : GetEscorts();
+	for(const weak_ptr<Ship> &ptr : allEscorts)
+	{
+		shared_ptr<Ship> escort = ptr.lock();
+		if(!escort || escort->IsParked() || escort->IsDestroyed() || GetSystem() != escort->GetSystem() || (!escort->IsYours() && !escort->GetPersonality().IsEscort()))
+			continue;
+		shared_ptr<const Ship> escortTarget = escort->GetTargetShip();
+		if(escortTarget)
+		{
+			if(gov->IsEnemy(escortTarget->GetGovernment()))
+				return true;
+		}
+	}
+	return false;
+}
+
+
+
+// If a ship is nearly out of battery energy and is slow to charge or no charge,
+// then consider it low on energy.
+bool Ship::IsEnergyLow() const
+{
+	double frames = 60.;
+	double idleEnergy = frames * GetIdleEnergyPerFrame();
+	double maxEnergy = Attributes().Get("energy capacity");
+	double totalConsumption = frames * GetEnergyConsumptionPerFrame();
+	double currentEnergy = (CanBeCarried() && GetSystem() && !landingPlanet) ? GetCurrentEnergy() : maxEnergy * .75;
+	//double currentEnergy = GetCurrentEnergy();
+	if(totalConsumption < 0.)
+	{
+		double secondsToEmpty = currentEnergy / (-totalConsumption);
+		// how quickly can it charge under no energy load
+		double secondsToFullCharge = (idleEnergy > 0.) ? (maxEnergy - currentEnergy) / idleEnergy : minimumOperatingTime + 1.;
+		if((secondsToEmpty < minimumOperatingTime && secondsToFullCharge > minimumOperatingTime) || (secondsToEmpty < minimumOperatingTime && idleEnergy <= 0.))
+			return true;
+	}
+	return false;
+}
+
+
+
+// Check if escort fleet is full of fuel.  In this case, the fighters can help
+// replenish the Carrier fuel tanker.
+bool Ship::IsEscortsFullOfFuel() const
+{
+	if(!IsYours())
+		return false;
+	const std::vector<std::weak_ptr<Ship>> allEscorts = (CanBeCarried() && GetParent()) ? GetParent()->GetEscorts() : GetEscorts();
+	for(const weak_ptr<Ship> &ptr : allEscorts)
+	{
+		shared_ptr<Ship> escort = ptr.lock();
+		if(!escort || escort->IsParked() || escort->IsDestroyed() || GetSystem() != escort->GetSystem() || (!escort->IsYours() && !escort->GetPersonality().IsEscort()))
```

Comment:
This is referenced a couple of places.  Should I DRY this up with a method?  Example header:

```cpp
bool IsEscortedBy(Ship &other);
```

Which returns true or false implemented based on this conditional.

Reply:
Yeah sounds good, but as a function defined inside Ship.cpp at the top inside the anonymous namespace, not as a member function.

Reply:
Ok will do thanks for guidance